### ISSUE TITLE
Potential race condition in nodemanager's initialize()

### ISSUE THREAD

**Issue by mikeyk** [alisaifee/coredis#106]
Hello!

We're seeing some odd errors in prod where `Redis Cluster cannot be connected. Please provide at least one reachable node.` is being thrown, but the initialization has actually completed. The only reason we can find that this could happen is:

* A new Python process is started
* Two coroutines both use the Redis cluster
* The first one starts initializing the cluster...
* ...but the second one enters the `initialize()` code as well, setting the `startup_nodes_reachable` flag to False
* Then when the first goes to check the flag (which it had set to True), it's now False and cluster initialization fails

After this first concurrent use of the cluster, all other accesses in that process work fine since the cluster is initialized.

Perhaps an `asyncio.Lock` around the initialization code? Happy to put up a Pull Request if that sounds like a reasonable approach.

**Comment by alisaifee** [comment]
Hi there!

From looking at the code there are definitely issues with concurrent initialization of the cluster client and perhaps worse than the error, you can also get into a situation where multiple co-routines send over the `CLUSTER SLOTS` command to initialize the node mapping when [ClusterConnectionPool.initialized=False](https://github.com/alisaifee/coredis/blob/master/coredis/pool/cluster.py#L134). 

I was able to (inconsistently) reproduce the `Please provide at least one reachable node` error with this example:
```python
from __future__ import annotations

import asyncio
import coredis


async def task(redis):
    await redis.ping()


async def run(n=15000):
    pool = coredis.BlockingClusterConnectionPool.from_url(
        url="redis://localhost:7000" max_connections=64
    )
    redis = coredis.RedisCluster(connection_pool=pool)
    tasks = [asyncio.create_task(task(redis)) for _ in range(n)]
    await asyncio.gather(*tasks)


if __name__ == "__main__":
    asyncio.run(run(n=15000))
```

The issue can be made to go away if the initialization of the client is done with an await, for example:
```python
redis = await RedisCluster(....)
```

instead of 
```python
redis = RedisCluster(...)
```

Having said that, a lock is still needed since it isn't a requirement to initialize the clients in an async context - so TL;DR: happy to take a PR!

**Comment by alisaifee** [comment]
@mikeyk just curious, does using the "async initialization" solution work for you?

**Comment by mikeyk** [comment]
> @mikeyk just curious, does using the "async initialization" solution work for you?

Thanks @alisaifee will try the async initialize today!

**Comment by alisaifee** [comment]
Not released yet, but 29dc71970131deed0948132eb1b4d65b9fb2aba6 should address the concurrency issue in the case where the client was initialized in a synchronous context.

**Comment by alisaifee** [comment]
[4.6.0](https://github.com/alisaifee/coredis/releases/tag/4.6.0) contains the synchronized initialization for the node manager in the cluster client (in case the async initialization isn't possible for your use case). I'll leave this open till you've had a chance to verify.

**Comment by mikeyk** [comment]
Hi @alisaifee! Verified fix in prod with 4.6.0, many thanks
